File: eyan_helper/src/main/java/eu/eyan/util/io/FileLineOffsetsReader.java
package eu.eyan.util.io;

import java.io.Closeable;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

import eu.eyan.util.tuple.Tuple2;

public class FileLineOffsetsReader implements Closeable {
	private FileReader reader;
	private final char[] buffer;
	int bufferOffset = 0;
	int bufferLength = 0;
	Character lastChar = '_';
	long nextLineLength = 0L;
	boolean finished = true;
	private long offset = 0;

	public FileLineOffsetsReader(String path, int bufferSize) throws FileNotFoundException {
		buffer = new char[bufferSize];
		reader = new FileReader(path);
		finished = false;
		offset = 0;
	}

	public FileLineOffsetsReader(File file) throws FileNotFoundException {
		buffer = new char[8192];
		reader = new FileReader(file);
		finished = false;
		offset = 0;
	}

	public Tuple2<Long, Long> readLine() throws IOException {
        while (true)
        {
            Character nextChar = nextChar();
            if (nextChar == null)
            {
                if (nextLineLength == 0)
                {
                    return null;
                }
                else
                {
                	Tuple2<Long, Long> ret = Tuple2.of(offset, offset+nextLineLength-1);
                	offset += nextLineLength;
                	nextLineLength=0;
            		return ret;
                }
            }
            else
            {
                if (lf(nextChar))
                {
            		nextLineLength+=1;
            		lastChar = null;
                	Tuple2<Long, Long> ret = Tuple2.of(offset, offset+nextLineLength-1);
                	offset += nextLineLength;
                	nextLineLength=0;
            		return ret;
                }
                else if (cr(lastChar))
                {
                	Tuple2<Long, Long> ret = Tuple2.of(offset, offset+nextLineLength-1);
                	offset += nextLineLength;
            		nextLineLength = 1;
            		lastChar = nextChar;
            		return ret;
                }
                else
                {
                	nextLineLength+=1;
            		lastChar = nextChar;
                }
            }
        }
    }

	private Character nextChar() throws IOException {
		if (finished) {
			return null;
		} else if (bufferOffset == bufferLength) {
			bufferLength = reader.read(buffer);
			bufferOffset = 0;
			if (bufferLength == -1) {
				finished = true;
				return null;
			}
		}
		char c = buffer[bufferOffset++];
		return c;
	}

	private boolean cr(Character c) {
		return c != null && c == '\r';
	}

	private boolean lf(Character c) {
		return c != null && c == '\n';
	}

	public long getOffset() {
		return offset;
	}

	@Override
	public void close() throws IOException {
		if (reader != null) {
			reader.close();
		}
	}

}

File: eyan_helper/src/main/scala/eu/eyan/util/io/FileLineOffsetsReader.scala
package eu.eyan.util.io

import java.io.File
import java.io.Closeable
import java.io.FileReader
import java.io.FileInputStream

object FileLineStartOffsetReader {
  def apply(file: File) = new FileLineStartOffsetReader(file.getAbsolutePath, 64 * 1024)
  def apply(file: String, bufferSize: Int) = new FileLineStartOffsetReader(file, bufferSize)
}

class FileLineStartOffsetReader private (file: String, bufferSize: Int) extends Closeable {
  val EGAL = 'x'
  private var offset = 0L

  private val byteBuffer = new Array[Byte](bufferSize)
  private val fis = new FileInputStream(file)

  private var lastChar = EGAL
  private var bufferOffset = 0
  private var bufferLength = 0
  private var byteBufferLength = 0
  private var nextLineLength = 0L

  def getOffset = offset

  override def close() = { fis.close() }

  /**
   * @return start offset from the line. After calling this method the end
   *         offset of the line can be read out with the getOffset():long
   * @throws IOException
   */
  def readLine = {
    var ready = false
    val INVALID_RESULT = -11L
    var result = INVALID_RESULT
    while (result < -1L) {
      val nextCharInt = nextChar
      if (nextCharInt == -1)
        if (nextLineLength == 0) result = -1
        else {
          val ret = offset
          offset += nextLineLength
          nextLineLength = 0
          result = ret
        }
      else {
        val nextChar = nextCharInt.toChar
        if (nextChar == '\n') {
          nextLineLength += 1
          lastChar = EGAL
          val ret = offset
          offset += nextLineLength
          nextLineLength = 0
          result = ret
        }
        else if (lastChar == '\r') {
          val ret = offset
          offset += nextLineLength
          nextLineLength = 1
          lastChar = nextChar
          result = ret
        }
        else {
          nextLineLength += 1
          lastChar = nextChar
        }
      }
    }
    result
  }

  private def nextChar: Int = {
    readFromFileToBufferIfNecessary
    if (byteBufferLength != -1) nextCharFromBuffer
    else -1
  }

  private def readFromFileToBufferIfNecessary =
    if (bufferOffset == byteBufferLength) {
      byteBufferLength = fis.read(byteBuffer)
      bufferOffset = 0
    }

  private def nextCharFromBuffer = { val c = byteBuffer(bufferOffset); bufferOffset += 1; c }
}